### PR TITLE
LORE-95: Develop Implementation for Transformation API

### PR DESCRIPTION
**Pull request by figliold**
Initial Implementation of the Transformation API

@mcalcote 
@cjlange 
@josephthweatt

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+    <properties>
+        <codice.scm.connection.url/>
+        <snapshots.repository.url />
+        <releases.repository.url />
+        <reports.repository.url/>
+        <maven-checkstyle-plugin.version>3.0.0</maven-checkstyle-plugin.version>
+
+        <!-- Build properties -->
+        <java.version>8</java.version>
```
Comment (mcalcote) [inline pom.xml]: ✏️ 11?

**Comment by brendan-hofmann** [comment]
build now

**Comment by brendan-hofmann** [comment]
We need a Jenkinsfile.

File: pom.xml (hunk 1 above)
Comment (figliold) [reply pom.xml]: Changed to Java 11

File: distros/docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# TODO - Replace with an openjdk/alpine base image if possible
+FROM adoptopenjdk/openjdk11
+VOLUME /tmp
+ARG JAR_FILE
+COPY ${JAR_FILE} app.jar
+# Turns on JMX to expose metrics
+# TODO - Need to automatically set java.rmi.server.hostname
```
[the remark this answers is not in the document]
Comment (figliold) [reply distros/docker/Dockerfile]: Added

File: distros/spring/pom.xml [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.springfox</groupId>
+            <artifactId>springfox-bean-validators</artifactId>
+            <version>${springfox.version}</version>
+        </dependency>
+
+        <!-- Test Dependencies
+             JUnit, Spring Test, Spring Boot Test, Mockito, AssertJ, Hamcrest, JSONassert and
+             JsonPath -->
```
[the remark this answers is not in the document]
Comment (figliold) [reply distros/spring/pom.xml]: Removed

File: distros/spring/src/main/java/com/connexta/transformation/service/Application.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
+import org.springframework.web.filter.CommonsRequestLoggingFilter;
+
+@SpringBootApplication
+public class Application {
```
[the remark this answers is not in the document]
Comment (figliold) [reply distros/spring/src/main/java/com/connexta/transformation/service/Application.java]: added

File: distros/spring/src/main/java/com/connexta/transformation/service/TransformController.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.web.bind.annotation.CrossOrigin;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@CrossOrigin(origins = "*")
+public class TransformController implements TransformApi {
```
[the remark this answers is not in the document]
Comment (figliold) [reply distros/spring/src/main/java/com/connexta/transformation/service/TransformController.java]: added

File: distros/spring/src/main/java/com/connexta/transformation/service/TransformController.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    this.transformationService = transformationService;
+  }
+
+  @RequestMapping(
+      value = "/transform",
+      produces = {MediaType.APPLICATION_JSON_VALUE},
+      consumes = {MediaType.APPLICATION_JSON_VALUE},
+      method = RequestMethod.POST)
+  @Override
+  public ResponseEntity<TransformResponse> transform(
```
[the remark this answers is not in the document]
Comment (figliold) [reply distros/spring/src/main/java/com/connexta/transformation/service/TransformController.java]: added

File: impl/src/main/java/com/connexta/transformation/service/TransformationServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.amqp.AmqpIOException;
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.rabbit.connection.CorrelationData;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.stereotype.Service;
+
+@Service
+public class TransformationServiceImpl implements TransformationService {
```
[the remark this answers is not in the document]
Comment (figliold) [reply impl/src/main/java/com/connexta/transformation/service/TransformationServiceImpl.java]: added

File: impl/src/main/java/com/connexta/transformation/service/TransformationServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  private static final String TRANSFORM_RESPONSE_TEMPLATE =
+      "The request with ID %s has been accepted for processing.";
+
+  private final String exchange;
+
+  private final String routingKey;
+
+  private final RabbitTemplate rabbitTemplate;
+
+  public TransformationServiceImpl(
```
[the remark this answers is not in the document]
Comment (figliold) [reply impl/src/main/java/com/connexta/transformation/service/TransformationServiceImpl.java]: added

File: impl/src/main/java/com/connexta/transformation/service/TransformationServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    Preconditions.checkNotNull(rabbitTemplate, "Rabbit Template cannot be null.");
+    Preconditions.checkNotNull(exchange, "Exchange cannot be null.");
+    Preconditions.checkNotNull(routingKey, "Routing Key cannot be null.");
+    this.rabbitTemplate = rabbitTemplate;
+    this.exchange = exchange;
+    this.routingKey = routingKey;
+  }
+
+  @Override
+  public TransformResponse transform(TransformRequest transformRequest) {
```
[the remark this answers is not in the document]
Comment (figliold) [reply impl/src/main/java/com/connexta/transformation/service/TransformationServiceImpl.java]: added

File: impl/src/main/java/com/connexta/transformation/service/TransformationServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    try {
+      forwardTransformRequest(transformRequest, new CorrelationData(transformRequest.getId()));
+    } catch (AmqpIOException e) {
+      failRequest(transformRequest, e);
+    }
+    return createTransformResponse(transformRequest);
+  }
+
+  private void forwardTransformRequest(
```
[the remark this answers is not in the document]
Comment (figliold) [reply impl/src/main/java/com/connexta/transformation/service/TransformationServiceImpl.java]: added

File: impl/src/main/java/com/connexta/transformation/service/TransformationServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+      TransformRequest transformRequest, CorrelationData correlationData) {
+    rabbitTemplate.convertAndSend(exchange, routingKey, transformRequest, correlationData);
+  }
+
+  private void failRequest(TransformRequest transformRequest, Exception e) {
+    // Should we store failed requests for later processing, or should return an error response back
+    // to the
+    // caller stating that the request failed processing?
```
[the remark this answers is not in the document]
Comment (figliold) [reply impl/src/main/java/com/connexta/transformation/service/TransformationServiceImpl.java]: added

**Pull request merged**